src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,529 @@
+            WriteParseOptions(writer, syntaxTree.Options);
+            writer.WriteObjectEnd();
+        }
+
+        private void WriteSourceText(JsonWriter writer, SourceText? sourceText)
+        {
+            if (sourceText is null)
+            {
+                return;
+            }
+
+            writer.WriteObjectStart();
+            WriteByteArrayValue(writer, "checksum", sourceText.GetChecksum().AsSpan());
+            writer.Write("checksumAlgorithm", sourceText.ChecksumAlgorithm);
+            writer.Write("encoding", sourceText.Encoding?.EncodingName);
+            writer.WriteObjectEnd();
+        }
+
+        internal void WriteMetadataReference(
+            JsonWriter writer,
+            MetadataReference reference,
+            DeterministicKeyOptions deterministicKeyOptions,
+            CancellationToken cancellationToken)
+        {
+            writer.WriteObjectStart();
+            if (reference is PortableExecutableReference peReference)
+            {
+                switch (peReference.GetMetadata())
+                {
+                    case AssemblyMetadata assemblyMetadata:
+                        {
+                            var modules = assemblyMetadata.GetModules();
+                            writeModuleMetadata(modules[0]);
+                            writer.WriteKey("secondaryModules");
+                            writer.WriteArrayStart();
+                            for (var i = 1; i < modules.Length; i++)
+                            {
+                                writer.WriteObjectStart();
+                                writeModuleMetadata(modules[i]);
+                                writer.WriteObjectEnd();
+                            }
+                            writer.WriteArrayEnd();
+                        }
+                        break;
+                    case ModuleMetadata m:
+                        writeModuleMetadata(m);
+                        break;
+                    default:
+                        throw new NotSupportedException();
+                }
+
+                writer.WriteKey("properties");
+                writeMetadataReferenceProperties(writer, reference.Properties);
+
+                void writeModuleMetadata(ModuleMetadata moduleMetadata)
+                {
+                    // The path of a reference, unlike the path of a file, does not contribute to the output
+                    // of the compilation. Only the MVID, name and version contribute here hence the file path
+                    // is deliberately omitted here.
+                    var peReader = moduleMetadata.GetMetadataReader();
+                    if (peReader.IsAssembly)
+                    {
+                        var assemblyDef = peReader.GetAssemblyDefinition();
+                        writer.Write("name", peReader.GetString(assemblyDef.Name));
+                        writer.Write("version", assemblyDef.Version.ToString());
```

Comment:
nit: audit tostring.

Reply:
This `ToString` is maddening because the .NET Core version has an `IFormattable` implementation I can use while desktop does not. Further the numbers are all small value `int` which should be naturally invariant but still want to be sure.